### PR TITLE
Support for Transparent Proxy

### PR DESCRIPTION
**Pull request by abbasvalliani**
Features include:

1) Added support for proxy to function as "transparent"
2) Generation of certificates compliant with new Mac OS certificate requirements
3) Added Mac OS ext cert support for new cert generation

### CODE REVIEW

**Review by chhsiao90 — CHANGES_REQUESTED** [review]
Thanks for the pull request. All looks great to me. I leave few comments, please take a look.

File: pom.xml [hunk 1]
```diff
@@ -1,202 +1,239 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
-  <modelVersion>4.0.0</modelVersion>
```
[the remark this answers is not in the document]
Comment (abbasvalliani) [reply pom.xml:3]: Will do. Changes were mainly in how the .sh was calling mvn to fork a separate process which you have to kill every time. I refactored to allow both inprocess and a forked execution.

File: src/main/java/com/github/chhsiao90/nitmproxy/handler/protocol/http1/Http1FrontendHandler.java [hunk 2]
```diff
@@ -133,4 +137,36 @@ private void handleHttpProxyConnection(ChannelHandlerContext ctx,
             connectionContext.tlsCtx().protocolPromise().setSuccess(Protocols.HTTP_1);
         }
     }
+
+    private void handleTransparentProxyConnection(ChannelHandlerContext ctx,
+                                                  FullHttpRequest request) throws Exception {
+        String hostHeader = request.headers().get(HttpHeaderNames.HOST);
+        String[] hostPort = hostHeader.split(":");
```
[the remark this answers is not in the document]
Comment (abbasvalliani) [reply src/main/java/com/github/chhsiao90/nitmproxy/handler/protocol/http1/Http1FrontendHandler.java:144]: Just to make sure I understand, do you prefer to resolve the DNS to an IP here? It does get resolved in the BackendChannelBootstrap.

File: src/main/java/com/github/chhsiao90/nitmproxy/handler/protocol/tls/AbstractAlpnHandler.java [hunk 3]
```diff
@@ -119,7 +122,7 @@ protected void onLookupComplete(ChannelHandlerContext ctx, Future<T> future) thr
     }
 
     protected abstract void onLookupComplete(ChannelHandlerContext ctx,
-                                             List<String> hostname, Future<T> future) throws Exception;
+                                             List<String> protocols, Future<T> future) throws Exception;
```
[the remark this answers is not in the document]
Comment (abbasvalliani) [reply src/main/java/com/github/chhsiao90/nitmproxy/handler/protocol/tls/AbstractAlpnHandler.java:125]: No worries. It was a minor typo!

File: src/main/java/com/github/chhsiao90/nitmproxy/handler/proxy/TransparentProxyHandler.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+    private void configHttp1(ChannelHandlerContext ctx) {
+        ctx.pipeline().replace(this, null, connectionContext.provider().http1FrontendHandler());
+    }
+
+    private void configHttp2(ChannelHandlerContext ctx) {
+        ctx.pipeline().replace(this, null, connectionContext.provider().http2FrontendHandler());
+    }
+
+    private class DetectSslHandler extends SslClientHelloHandler<Boolean> {
```
[the remark this answers is not in the document]
Comment (abbasvalliani) [reply src/main/java/com/github/chhsiao90/nitmproxy/handler/proxy/TransparentProxyHandler.java:88]: Will take a deeper look on refactoring it today.

**Comment by abbasvalliani** [comment]
> Subtract one year from the current time for the "from" field of the cert validity period, don't do from now. Minor issues in clock synchronization can cause some sessions to reject the certificate.

Good point. Clock sync becomes a major issue. Will resolve it in the next PR.

File: src/main/java/com/github/chhsiao90/nitmproxy/handler/protocol/http1/Http1FrontendHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chhsiao90) [reply src/main/java/com/github/chhsiao90/nitmproxy/handler/protocol/http1/Http1FrontendHandler.java:144]: I mean [Address.resolve](https://github.com/chhsiao90/nitmproxy/blob/master/src/main/java/com/github/chhsiao90/nitmproxy/Address.java#L25-L32), I guessed we can reuse this class to parse the host.

File: src/main/java/com/github/chhsiao90/nitmproxy/handler/proxy/TransparentProxyHandler.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+public class TransparentProxyHandler extends ChannelHandlerAdapter {
+    private ConnectionContext connectionContext;
+
+    public TransparentProxyHandler(ConnectionContext connectionContext) {
+        this.connectionContext = connectionContext;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.pipeline().replace(this, null, connectionContext.provider().tlsFrontendHandler());
```
Comment (chhsiao90) [inline src/main/java/com/github/chhsiao90/nitmproxy/handler/proxy/TransparentProxyHandler.java:16]: Thanks for addressing the comment, it looks great now!

**Review by chhsiao90 — APPROVED** [review]

**Comment by chhsiao90** [comment]
Thanks @abbasvalliani for the great work!

**Pull request merged**